### DISCUSSION TITLE
Github checkout action doesn't see PR modifications

### DISCUSSION THREAD

**Discussion by TomerFi** [community/community#22363] [New to GitHub]
<h2>
<a name="hello-everyone-1" class="anchor" href="#hello-everyone-1"></a>Hello everyone</h2>
<p>I have an issue I’ve been tracking for a long time and I was never able to quite pinpoint it, until today.<br>
It appears that for pull requests, modifications are not always grabbed by the checkout action.</p>
<p>I’ll use the example I stumbled upon today to clarify myself.</p>
<p>If you take a look at the following PR, created by dependabot:<br>
<a href="https://github.com/TomerFi/switcher_webapi/pull/407" class="inline-onebox" rel="noopener nofollow ugc">build(deps): bump aiohttp from 3.7.4.post0 to 3.8.1 by dependabot[bot] · Pull Request #407 · TomerFi/switcher_webapi · GitHub</a>.</p>
<p>You will note in the <a href="https://github.com/TomerFi/switcher_webapi/pull/407/files" rel="noopener nofollow ugc">files</a> tab that the only modification introduced by this PR is modifying the <code>requirements.txt</code>, bumping the <code>aiohttp</code> requirement from <code>3.7.4.post0</code> to <code>3.8.1</code>.</p>
<p>Now if you take a look at the following workflow run, triggered by the above PR:<br>
<a href="https://github.com/TomerFi/switcher_webapi/runs/4440694337?check_suite_focus=true" class="inline-onebox" rel="noopener nofollow ugc">build(deps): bump aiohttp from 3.7.4.post0 to 3.8.1 · TomerFi/switcher_webapi@858d303 · GitHub</a>.</p>
<p>You’ll see I got 4 jobs in this workflow, it doesn’t matter what they do for this particular issue of mine,<br>
just note the first two steps of each, they were put there to showcase my point.</p>
<h2>
<a name="the-lint-documentation-files-job-2" class="anchor" href="#the-lint-documentation-files-job-2"></a>The <code>Lint documentation files</code> job</h2>
<p>The first two steps are:</p>
<pre><code class="lang-yaml">    steps:
      - name: Checkout sources
        uses: actions/checkout@v2.4.0

      - name: temp pring requirements.txt
        run: cat requirements.txt
</code></pre>
<p>The contents of the <code>requirements.txt</code>:</p>
<pre><code class="lang-auto">aioswitcher==2.0.7
aiohttp==3.7.4.post0
</code></pre>
<h2>
<a name="the-build-project-job-3" class="anchor" href="#the-build-project-job-3"></a>The <code>Build project</code> job</h2>
<p>The first two steps are:</p>
<pre><code class="lang-yaml">    steps:
      - name: Checkout sources
        uses: actions/checkout@v2.4.0
        with:
          ref: ${{ github.head_ref }}

      - name: temp pring requirements.txt
        run: cat requirements.txt
</code></pre>
<p>The contents of the <code>requirements.txt</code>:</p>
<pre><code class="lang-auto">aioswitcher==2.0.7
aiohttp==3.8.1
</code></pre>
<h2>
<a name="the-build-docker-job-4" class="anchor" href="#the-build-docker-job-4"></a>The <code>Build docker</code> job</h2>
<p>The first two steps are:</p>
<pre><code class="lang-yaml">    steps:
      - name: Checkout sources
        uses: actions/checkout@v2.4.0
        with:
          ref: ${{ github.sha }}

      - name: temp pring requirements.txt
        run: cat requirements.txt
</code></pre>
<p>The contents of the <code>requirements.txt</code>:</p>
<pre><code class="lang-auto">aioswitcher==2.0.7
aiohttp==3.7.4.post0
</code></pre>
<h2>
<a name="the-push-coverage-report-to-codecov-job-5" class="anchor" href="#the-push-coverage-report-to-codecov-job-5"></a>The <code>Push coverage report to CodeCov</code> job</h2>
<p>The first two steps are:</p>
<pre><code class="lang-yaml">    steps:
      - name: Source checkout
        uses: actions/checkout@v2.4.0
        with:
          fetch_depth: 0

      - name: temp pring requirements.txt
        run: cat requirements.txt
</code></pre>
<p>The contents of the <code>requirements.txt</code>:</p>
<pre><code class="lang-auto">aioswitcher==2.0.7
aiohttp==3.7.4.post0
</code></pre>
<p>It would appear that if I need the PR to see the correct modifications,<br>
I need to manually configure the <code>checkout action</code> to checkout the branch from which the PR was created.<br>
All other checkouts, actually checkout the merge commit id, and for some weird reason, don’t see the modifications from the PR.</p>
<p>This caused me, on a couple of occasions, to merge a PR just to eventually see it fail other workflows triggered by the merge, and eventually, I had to manually do a <code>git restore</code> and force push it for each occurrence.</p>
<p>Am I missing something here?</p>

**Comment by TomerFi** [comment, marked as answer]
<p>Figured it out!</p>
<p>The <a href="https://github.com/actions/checkout#usage" rel="noopener nofollow ugc">checkout action</a> uses the <code>sha</code> value for the checkout process by default.<br>
And I was using <code>pull_request_target</code> instead of <code>pull_request</code> as the trigger for the workflow.</p>
<p>Based on the documentation:<br>
When using <a href="https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#pull_request" rel="noopener nofollow ugc">pull_request</a> as the trigger:</p>
<ul>
<li>GITHUB_SHA = Last merge commit on the <code>GITHUB_REF</code> branch</li>
<li>GITHUB_REF = PR merge branch <code>refs/pull/:prNumber/merge</code>
</li>
</ul>
<p>When using <a href="https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#pull_request" rel="noopener nofollow ugc">pull_request_target</a> as the trigger:</p>
<ul>
<li>GITHUB_SHA = Last commit on the PR base branch</li>
<li>GITHUB_REF = PR base branch</li>
</ul>
<p>So I was actually checking out the base branch because I was using <code>pull_request_target</code>.</p>
<p>As a workaround, if you want to use <code>pull_request_target</code> as the trigger, you can configure the <code>checkout action</code> to use the <code>head_ref</code> in order to grab the PR base branch:</p>
<pre><code class="lang-yaml">    steps:
      - name: Checkout sources
        uses: actions/checkout@v2
        with:
          ref: ${{ github.head_ref }}
</code></pre>

**Comment by jsoref** [comment]
<h1>
<a name="here-there-be-dragons-1" class="anchor" href="#here-there-be-dragons-1"></a>Here there be :dragon:s</h1>
<p>Be <strong>very</strong> careful about using the head ref from a <code>pull_request_target</code>.</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/2119212/181070548-b4f24ac7-560b-4250-9814-30bffa5fd7a2.png" class="site-icon" width="72" height="72">

      <a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" target="_blank" rel="noopener nofollow ugc" title="12:00AM - 15 December 2020">GitHub Security Lab – 15 Dec 20</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://user-images.githubusercontent.com/2119212/181070549-efc18caf-01c6-4ea5-9269-2fa9bc222781.png" class="thumbnail" width="690" height="362" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_690x362.png, https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_1035x543.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_10x10.png"></div>

<h3><a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" target="_blank" rel="noopener nofollow ugc">Keeping your GitHub Actions and workflows secure Part 1: Preventing pwn requests</a></h3>

  <p>In this article, we’ll discuss some common security malpractices for GitHub Actions and workflows, and how to best avoid them. Our examples are based on real-world GitHub workflow implementation vulnerabilities the GitHub Security Lab has reported to...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<h1>
<a name="races-and-conflicts-2" class="anchor" href="#races-and-conflicts-2"></a>Races :racehorse: and Conflicts :crossed_swords:
</h1>
<p>Fwiw, I’ve just finished debugging the fact that <code>refs/pull/:prNumber/merge</code> is unreliable – it’s apparently generated by an asynchronous task that afaict doesn’t properly handle merge conflicts after an update. In my experience that meant that if the latest pushed head conflicted with the base branch, the <code>refs/pull/:prNumber/merge</code> commit would still exist, but just point to the previous merge result even though the PR itself would correctly say that the PR couldn’t be merged because there’s a conflict.</p>

**Comment by TomerFi** [comment]
<aside class="quote no-group quote-modified" data-username="jsoref" data-post="3" data-topic="216132">
<div class="title">
<div class="quote-controls"></div>
 jsoref:</div>
<blockquote>
<p>Be <strong>very</strong> careful about using the head ref from a <code>pull_request_target</code> .</p>
<p><a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" class="inline-onebox" rel="noopener nofollow ugc">Keeping your GitHub Actions and workflows secure Part 1: Preventing pwn requests | GitHub Security Lab</a></p>
</blockquote>
</aside>
<p>This is an excellent post, thank you very much for sharing,<br>
I’m definitely considering modifications to my workflows after reading it.</p>
<p>I should point out that I do use a couple of other safety precautions:</p>
<ul>
<li>
<p>I use <a href="https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token" rel="noopener nofollow ugc">permissions for the GITHUB_TOKEN</a> to limit the access for workflows triggered by <code>pull_request_target</code>.</p>
</li>
<li>
<p>I configure my repositories to always require my approval for all outside collaborators and not just first-time ones:<br>
<img src="https://user-images.githubusercontent.com/28388442/181070551-8977fd64-8ac1-4f58-a547-cb13df1b7df2.png" alt="image" data-base62-sha1="2RrYjvDJUa9tOGDldMb7XeE1Y22" width="512" height="227"></p>
</li>
</ul>
<aside class="quote no-group" data-username="jsoref" data-post="3" data-topic="216132">
<div class="title">
<div class="quote-controls"></div>
 jsoref:</div>
<blockquote>
<p>Fwiw, I’ve just finished debugging the fact that <code>refs/pull/:prNumber/merge</code> is unreliable – it’s apparently generated by an asynchronous task that afaict doesn’t properly handle merge conflicts after an update. In my experience that meant that if the latest pushed head conflicted with the base branch, the <code>refs/pull/:prNumber/merge</code> commit would still exist, but just point to the previous merge result even though the PR itself would correctly say that the PR couldn’t be merged because there’s a conflict.</p>
</blockquote>
</aside>
<p>This is interesting, I’d love to hear more about this.</p>

**Comment by jsoref** [comment]
<p>I might write something up more formally (possibly on dev.to?), but for the time being, here’s what I have:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/check-spelling/checkout-merge">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/check-spelling/checkout-merge" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181070555-ffdcf75a-b62e-4e28-91d8-09bf26258434.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/4/a/4a69d46bfec6b02cea18ebc7295328e07a7a456b_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/4/a/4a69d46bfec6b02cea18ebc7295328e07a7a456b_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/4/a/4a69d46bfec6b02cea18ebc7295328e07a7a456b.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/4/a/4a69d46bfec6b02cea18ebc7295328e07a7a456b_2_10x10.png"></div>

<h3><a href="https://github.com/check-spelling/checkout-merge" target="_blank" rel="noopener nofollow ugc">GitHub - check-spelling/checkout-merge: Action for synthesizing a merge commit</a></h3>

  <p>Action for synthesizing a merge commit. Contribute to check-spelling/checkout-merge development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>I’ve basically replaced all instances where I ask for the magic commit with this action that generates it on the fly. In terms of cost, it’s a bit of extra cpu time (and unfortunately a lot of git history fetching since I don’t know offhand of an easy way to find the greatest common ancestor using GitHub APIs and didn’t want to spend the brain power &amp; cpu power to find it for something that isn’t particularly important on average).</p>
<p>I’m not done deploying this change internally but I expect to have it fully deployed internally by the end of the month and then as I do a refresh of my check-spelling workflow, I’ll include this as well.</p>
